packages/slice-machine/src/hooks/useLocalStorageItem.ts

```diff
@@ -0,0 +1,62 @@
+import { Dispatch, SetStateAction, useMemo, useSyncExternalStore } from "react";
+
+export type UseLocalStorageItemInfo = {
+  /** Whether the item is unset in localStorage and using the defaultValue as a fallback. */
+  isUnset: boolean;
+};
+
+export type UseLocalStorageReturnType<T> = [
+  T,
+  Dispatch<SetStateAction<T>>,
+  UseLocalStorageItemInfo,
+];
+
+const SLICE_MACHINE_STORAGE_PREFIX = "slice-machine";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const isFunction = (value: any): value is (...args: any[]) => any => {
+  return typeof value === "function";
+};
+
+export function useLocalStorageItem<T>(
+  key: string,
+): UseLocalStorageReturnType<T | undefined>;
+export function useLocalStorageItem<T>(
+  key: string,
+  defaultValue: T,
+): UseLocalStorageReturnType<T>;
+export function useLocalStorageItem<T>(
+  key: string,
+  defaultValue?: T | undefined,
+): UseLocalStorageReturnType<T | undefined> {
+  const storageKey = `${SLICE_MACHINE_STORAGE_PREFIX}_${key}`;
+  const getSnapshot = () => localStorage.getItem(storageKey);
+
+  const subscribe = (listener: () => void) => {
+    window.addEventListener("storage", listener);
+    return () => void window.removeEventListener("storage", listener);
+  };
+
+  const serializedItem = useSyncExternalStore(subscribe, getSnapshot);
+  const item = useMemo(() => {
+    try {
+      if (serializedItem != null) return JSON.parse(serializedItem) as T;
+      return defaultValue;
+    } catch (error) {
+      console.warn(`Error reading localStorage key “${storageKey}”:`, error);
+      return defaultValue;
+    }
+  }, [serializedItem, defaultValue, storageKey]);
```

Comment:
My issue here, and the reason I went with a computed value here, was because I wasn't having success handling this logic inside `getSnapshot`, as it kept yelling at me with: `useLocalStorageItem.ts:43 Warning: The result of getSnapshot should be cached to avoid an infinite loop`, even with a try catch inside of it.

Reply:
You're right about those pitfalls, but I believe they can be solved by simply restricting how the hook can be used. Both `key` and `defaultValue` should be hardcoded or only the initial value is considered, just like `useState`/`useRef` initializer.

Reply:
No longer applicable. Resolving.